Add initial kernel module tutorial

NeisesMike opened the pull request:
Signed-off-by: Michael Neises <neisesmichael@gmail.com>

This merge would add a tutorial for building kernel modules for use with camkes-vm-linux. This tutorial differs from [this tutorial](https://docs.sel4.systems/Tutorials/camkes-vm-linux.html#adding-a-kernel-module) because this new procedure works in the case of simulation with qemu-arm-virt. This tutorial was motivated in part by the desire for rapid-prototyping, for which simulation is key.

This procedure has been tested to work using both Debian 10 and WSL2 on Windows 10.

This tutorial follows from [this thread](https://lists.sel4.systems/hyperkitty/list/devel@sel4.systems/thread/XOD6DXC33VRJWEATGIOP4YSGK37UKZYV/) and  [this thread](https://lists.sel4.systems/hyperkitty/list/devel@sel4.systems/thread/NOTNLAC7NS2VWZPBEA2M6CETSR5OOREV/?sort=date) on the devel mailing list.

NeisesMike commented:
I can see that I'm failing a "Links" check, and my best guess is this is because I have links to an external github page. I'm linking to my own repository which makes the process of using kernel modules in this way trivial. I'd be happy to add those files here, but as far as I can tell, no other tutorial adds such files.

Please advise on the best course of action.

lsf37 commented:
The links check fails on two unrelated links that should be up again in a few days (the website of the Trustworthy Systems group is moving), so that part is all fine.

It would be nice to include the files in this repo, so that they can be easily updated when things change over time. @kent-mcleod knows better how the doc-site build works, but I think it would be fine to include the files in the directory of this tutorial itself, and make the links point to the github repo here (seL4/sel4-tutorials).

Thanks for adding this, I'm sure this will help a lot of people!

NeisesMike commented:
I've added the example camkes-vm app to the tutorial directory here, and changed the corresponding links. I've peeked around a bit to try to figure out how the ["./init --tut"](https://docs.sel4.systems/Tutorials/#doing-the-tutorials) call operates, but I couldn't track it down. In lieu of that, there is still one link to the external repo, which just serves to make the build process more convenient. Please let me know if there's anything else I can do to smoothe out the merge process.

And it's my pleasure! I'd be thrilled if this helped someone else along.

lsf37 commented:
@NeisesMike sorry for the long pause -- could you please rebase the PR branch or enable write access for it so the checks can run?

NeisesMike commented:
@lsf37 No problem at all. I believe I've brought everything up-to-date.

nomadeel commented on tutorials/camkes-vm-kernel-module/camkes-vm-kernel-module.md:
````diff
@@ -0,0 +1,119 @@
+cp arch/arm64/boot/Image [...]/projects/camkes-vm-images/qemu-arm-virt/newLinux
+```
+
+6. edit your cmakelists.txt to use the new kernel by replacing this line:
````
There are multiple CMakeLists.txt files in the camkes VM project. I assume that this is referring to the tutorial application's CMakeLists.txt, but for someone new who isn't as familiar with the application structure they can be confused. Can you add something explicit here to point to the right file?

NeisesMike replied:
I've added greater detail here, and at all other steps. I would like to include a permalink to a line in an appropriate file (like the tutorials/camkes-vm-kernel-module/module_minimal/CMakeLists.txt file I've included here), but I'm not sure how to do that without the file already being hosted.

lsf37 replied:
That'd be fine -- if you can predict the github URL correctly (e.g. by example from another file in the repo), we can override the link checker for this. The link will become valid when the PR is merged.

nomadeel commented:
> I've added the example camkes-vm app to the tutorial directory here, and changed the corresponding links. I've peeked around a bit to try to figure out how the "./init --tut" call operates, but I couldn't track it down. 

The ./init script is actually this file [here](https://github.com/seL4/sel4-tutorials/blob/master/init.py). That then calls some functions [here](https://github.com/seL4/sel4-tutorials/blob/master/common.py), which then instantiates a Jinja session using [this](https://github.com/seL4/sel4-tutorials/blob/master/template.py) to take the tutorial markdown file and generate files for the tutorial.

lsf37 commented:
> I've failed the "style" and "whitespace" checks several times now, and I'm ready to cry uncle. I can't interpret what those checks want from me

The whitespace check flags trailing whitespace at the end of lines and at the end of files. It basically just runs `git diff master --check`, which you can replicate locally if you're not sure.

The style checks run the script `misc/style-all.sh` from the repo [sel4_tools](https://github.com/seL4/seL4_tools/tree/master/misc) on the files that are changed or added in this PR. The script figures out which language the file is and if there is a style checker configured for that language (mostly C, CMake, python and sh) will run that style checker. For C and python, the script will update the files such that they are style conforming, so you can run `misc/style-all.sh` locally to produce something that will pass.

If those two are the only missing bits, I'm also happy to run them and  fix up those parts for you. It looks like [this specific problem](https://github.com/seL4/sel4-tutorials/pull/61/checks?check_run_id=3761483996#step:2:27) actually crashes the style checker parser, which is pretty rare.

lsf37 commented:
(It looks like I can't actually fix them up for you, because I don't have permissions on your branch -- if you enable "allow maintainers push access", I'll be able to)

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/settings.cmake:
```diff
@@ -0,0 +1,18 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
I think what the CMake style checker is choking on is the comment style. This should work:
```suggestion
#
# Copyright 2021 Michael Neises
#
# SPDX-License-Identifier: BSD-2-Clause
#
```

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/CMakeLists.txt:
```diff
@@ -0,0 +1,117 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
```suggestion
#
# Copyright 2021 Michael Neises
#
# SPDX-License-Identifier: BSD-2-Clause
#
```

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/modules/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
```suggestion
#
# Copyright 2021 Michael Neises
#
# SPDX-License-Identifier: BSD-2-Clause
#
```

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/modules/hello/Makefile:
```diff
@@ -0,0 +1,10 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
Should use Makefile comment (`#`)

NeisesMike replied:
I've made the suggested edit.

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/modules/hello/hello.c:
```diff
@@ -0,0 +1,30 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
Should use C comments:
```suggestion
/*
 * Copyright 2021 Michael Neises
 *
  * SPDX-License-Identifier: BSD-2-Clause
  */
```

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/qemu-arm-virt/devices.camkes:
```diff
@@ -0,0 +1,45 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
Should use CAmkES comment (same as C style)

NeisesMike replied:
I've made the suggested edit.

lsf37 commented on tutorials/camkes-vm-kernel-module/module_minimal/module_minimal.camkes:
```diff
@@ -0,0 +1,30 @@
+<!--
+  Copyright 2021 Michael Neises
+
+  SPDX-License-Identifier: BSD-2-Clause
+-->
```
Should use C/CAmKES style comment

NeisesMike replied:
I've made the suggested edit.

NeisesMike commented:
> if you enable "allow maintainers push access"
I feel a bit sheepish, but I can't find this option despite looking all over for it.

I've made all the styling changes you've suggested, and with `git diff master --check` I was able to deduce that the whitespace issues were caused by my host system being Windows (inconsistent line endings). I was able to use the dos2unix program to resolve them.

lsf37 commented:
> I failed some checks for some github-generated commit messages that were too long, so I've tried to amend them here. I also failed a style check on one file, so I've tried to fix that as well.

Thanks for that. I must admit I also don't know where the button is for the maintainer-can-modify setting. I vaguely remember seeing it in the right bar, where the PR reviewers are, somewhere towards the top, but it only appears for the PR author. If you can't find and want it I can go looking a bit further and do a test PR myself.

kent-mcleod commented:
> I feel a bit sheepish, but I can't find this option despite looking all over for it.

If you are on a desktop then hopefully it can be found on the side panel underneath the **Lock Conversation** option:
- [ ] Allow edits and access to secrets by maintainers

NeisesMike commented:
I think I've confused the issue of "maintainer access" somewhat by doing this through a Github organization. I've invited Gerwin and Kent to be members with maintainer access to the repository in question. I think this is a solution, so I hope now to let you help me where I keep misunderstanding these style checks. Thank you again for your patience.

lsf37 commented:
Ok, I've done this now, merging all of the style/whitespace fixup commits into the most recent commits to the same files. I checked that the result is precisely the same text (empty diff) and that there are only commits with content changes left.

I would usually merge these even more if it was my own PR, but I think this is Ok to go ahead with.

lsf37 commented:
Will leave it open for a few more hours in case @kent-mcleod or @nomadeel still have any comment, but from my side this is ready to merge.

lsf37 commented:
Ok, all done. We should still follow up on whether it makes sense to integrate this into the `init` script, but that can be a separate PR.

Thanks for all your patience with the style checks!
